### PR TITLE
Modernize and prepare ESLint config for TypeScript

### PR DESCRIPTION
**Pull request by chriszarate**
## Proposal

- Drop `eslint-plugin-wpcalypso` in favor of `@wordpress/eslint-plugin`, which is actively maintained and includes helpful rules for TypeScript, React, and testing. The experience is quite nice.
- Move away from the painful `peerDependency` approach and add `@wordpress/eslint-plugin` as an actual dependency. Only `eslint` remains as a `peerDep`.
- Do not hardcode the base presets from `@wordpress/eslint-plugin` in this project. Instead, allow consumers to define their own base presets, which might vary from project to project.
- This project simply provides potential overrides, organized into its own presets (`typescript`, `cli`, `testing`, etc.).

## Example `eslintrc`s

```
{
	"extends": [
		"plugin:@wordpress/eslint-plugin/recommended",
		"plugin:wpvip/base",
		"plugin:wpvip/react",
		"plugin:wpvip/testing",
		"plugin:wpvip/typescript",
	]
}
```

or

```
{
	"extends": [
		"plugin:@wordpress/eslint-plugin/es5",
		"plugin:wpvip/base",
	]
}
```

or

```
{
	"extends": [
		"plugin:@wordpress/eslint-plugin/test-e2e",
		"plugin:wpvip/testing",
	]
}
```

etc.

## Feedback

Is this a good base for linting in 2022?

This copies over some existing rules from the main branch, but is otherwise a blank slate for us to build our TypeScript rulesets.

Note that there's a bug in the WordPress eslint-plugin related to `prettier` (which is supposed to be optional). PR here:
https://github.com/WordPress/gutenberg/pull/40634

### CODE REVIEW

File: __fixtures__/cli/disallowed.js [hunk 1]
```diff
@@ -0,0 +1 @@
+const unused = 1;
```
Comment (jblz) [inline __fixtures__/cli/disallowed.js]: Should this be moved to the base disallowed fixture file?  I don't think we want unused variables pretty much anywhere, right?

File: README.md [hunk 2]
```diff
@@ -1,41 +1,38 @@
...
-"eslintConfig": {
-  "extends": "wpvip"
+{
+	"extends": [
+		"plugin:@wordpress/eslint-plugin/recommended",
+		"plugin:wpvip/base",
+		"plugin:wpvip/react",
+		"plugin:wpvip/testing",
+		"plugin:wpvip/typescript",
+	]
```
Comment (jblz) [inline README.md:27]: Is there anything special that needs to be done in the extending config to use the `cli`-specific settings?

**Review by jblz — COMMENTED** [review]
This is awesome!

Left a couple of questions.

File: README.md (hunk 2 above)
Comment (chriszarate) [reply README.md:27]: No, they are just regular ESLint rules

Comment (chriszarate) [reply README.md:27]: The fixtures also serve to provide example `.eslintrc`s, you can see the CLI fixture here

https://github.com/Automattic/eslint-config-wpvip/pull/137/files#diff-5c70a1581c5e9c9930deca097d9e67a732adae44fe08af385427547bb5fe2682

**Comment by chriszarate** [comment]
New changes:
- Improved comments
- Improved fixtures
- Added `eslint-plugin-security`

File: __fixtures__/base/disallowed.js [hunk 3]
```diff
@@ -9,8 +9,20 @@ let things = {
 	f7: (  ) => ['x'],
 	12: '32',
 	ff: xx => xx,
-	gg: ( {fds, }) => 
+	gg: ( {fds, }) =>
         fds,
 };
 
-nonExistent();
```
Comment (jblz) [inline __fixtures__/base/disallowed.js]: The lack of a newline at the end of the `disallowed` fixture file was intentional (I thought it was a violation, but looks like it is not).

Comment (jblz) [reply __fixtures__/base/disallowed.js]: It is now :)

**Review by jblz — APPROVED** [review]
This is awesome, @chriszarate!

Thanks for working on it.

File: __fixtures__/base/disallowed.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+const cp = require( 'child_process' );
+
+const lol = 'rm -rf all_your_files';
+cp.exec( lol );
```
Comment (jblz) [inline __fixtures__/base/disallowed.js:4]: This line isn't getting marked as a warning or error in my IDE (here or in the cil/disallowed fixture).  Should it be?

Comment (chriszarate) [reply __fixtures__/base/disallowed.js:4]: Did you `npm i` to pick up the new security plugin?

Comment (jblz) [reply __fixtures__/base/disallowed.js:4]: Oh, whoops. I did when I started looking at this branch, but not after you added the security plugin. Don't mind me!

File: configs/typescript.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
+	extends: [
+		'plugin:@typescript-eslint/recommended',
+	],
+	/**
+	 * Please include a short description of the rule. For rules that downgrade or
+	 * disable errors, include a brief justification or reasoning.
+	 */
+	rules: {
```
Comment (saroshaga) [inline configs/typescript.js:9]: I know this is controversial but do we want to add `'@typescript-eslint/no-explicit-any': 'off',` here?

Comment (saroshaga) [reply configs/typescript.js:9]: Maybe even `@typescript-eslint/no-shadow`

Comment (saroshaga) [reply configs/typescript.js:9]: This is the eslint file we use for our NestJS services: https://github.com/Automattic/vip-site-details/blob/trunk/.eslintrc.js

**Review by saroshaga — COMMENTED** [review]
Thanks for working on this!! 

I think we start trying this out on our services.

File: configs/typescript.js (hunk 5 above)
Comment (chriszarate) [reply configs/typescript.js:9]: > I know this is controversial but

I think the answer to anything that feels controversial is to branch off into separate configs. Especially with TypeScript, we'll want to accommodate codebases that were born into TS vs codebases that are migrating. I can envision something like:

`typescript` <-- base recommended config
`typescript/migration` <-- overrides to relax things
`typescript/strict` <-- overrides to tighten things up

> This is the eslint file we use for our NestJS services

Thanks! This sounds like a great base for `typescript/strict`. I'll add it in a new PR.

Comment (saroshaga) [reply configs/typescript.js:9]: fair enough! Thanks!

**Pull request merged**